openverse_catalog/dags/data_refresh/dag_factory.py

```diff
@@ -1,135 +1,143 @@
-able to suspend itself and free the worker slot, because we want to lock the
-entire pool on waiting for a particular data refresh to run.
+) Importantly, the data refresh TaskGroup is also configured to handle concurrency
+requirements of the data refresh server.
 
 You can find more background information on this process in the following
 issues and related PRs:
 
 - [[Feature] Data refresh orchestration DAG](
 https://github.com/WordPress/openverse-catalog/issues/353)
+- [[Feature] Merge popularity calculations and data refresh into a single DAG](
+https://github.com/WordPress/openverse-catalog/issues/453)
 """
-import json
 import logging
-import os
 from typing import Sequence
-from urllib.parse import urlparse
 
 from airflow import DAG
-from airflow.exceptions import AirflowException
-from airflow.providers.http.operators.http import SimpleHttpOperator
-from airflow.providers.http.sensors.http import HttpSensor
-from common.constants import DAG_DEFAULT_ARGS, XCOM_PULL_TEMPLATE
-from common.sensors.single_run_external_dags_sensor import SingleRunExternalDAGsSensor
+from airflow.models.dagrun import DagRun
+from airflow.operators.python import BranchPythonOperator
+from airflow.settings import SASession
+from airflow.utils.session import provide_session
+from airflow.utils.state import State
+from common.constants import DAG_DEFAULT_ARGS
+from common.popularity import operators
+from data_refresh.data_refresh_task_factory import create_data_refresh_task_group
 from data_refresh.data_refresh_types import DATA_REFRESH_CONFIGS, DataRefresh
-from requests import Response
+from data_refresh.refresh_popularity_metrics_task_factory import (
+    GROUP_ID as REFRESH_POPULARITY_METRICS_GROUP_ID,
+)
+from data_refresh.refresh_popularity_metrics_task_factory import (
+    create_refresh_popularity_metrics_task_group,
+)
+from data_refresh.refresh_view_data_task_factory import create_refresh_view_data_task
 
 
 logger = logging.getLogger(__name__)
 
-
-DATA_REFRESH_POOL = "data_refresh"
+REFRESH_MATERIALIZED_VIEW_TASK_ID = operators.UPDATE_DB_VIEW_TASK_ID
+# The first task in the refresh_popularity_metrics TaskGroup
+REFRESH_POPULARITY_METRICS_TASK_ID = (
+    f"{REFRESH_POPULARITY_METRICS_GROUP_ID}"
+    f".{operators.UPDATE_MEDIA_POPULARITY_METRICS_TASK_ID}"
+)
 
 
-def response_filter_data_refresh(response: Response) -> str:
-    """
-    Filter for the `trigger_data_refresh` task, used to grab the endpoint needed
-    to poll for the status of the triggered data refresh. This information will
-    then be available via XCom in the downstream tasks.
+@provide_session
+def _month_check(dag_id: str, session: SASession = None) -> str:
     """
-    status_check_url = response.json()["status_check"]
-    return urlparse(status_check_url).path
+    Checks whether there has been a previous DagRun this month. If so,
+    returns the task_id for the matview refresh task; else, returns the
+    task_id for refresh popularity metrics task.
```

Comment:
Really good idea to check if there are any runs for the month, rather than doing a simple "if date.day == 1" check! ⭐ I was thinking the latter, this is so much better.